In [1]:
import pandas as pd
import datetime
import holidays

In [2]:
dati = pd.read_csv(r'Dati_PUN_17_19.csv', parse_dates = ['Data'])
sinusoidi = pd.read_csv(r'sinusoidi.csv')

In [3]:
dati['Giorno'] = dati['Data'].apply(lambda x: x.date().strftime('%A'))

In [4]:
iatly_holidays = holidays.Italy(years = [2017, 2018, 2019])

In [5]:
def get_holiday(row):
    row = row.date()
    iatly_holidays = holidays.Italy(years = [2017, 2018, 2019])
    if row in iatly_holidays:
        return(1)
    else:
        return(0)    

In [6]:
dati['Giorno_Vacanza'] = dati['Data'].apply(get_holiday)

In [7]:
dati.head()

Data    PUN  Giorno  Giorno_Vacanza
0 2017-01-01 00:00:00  53.30  Sunday               1
1 2017-01-01 01:00:00  52.00  Sunday               1
2 2017-01-01 02:00:00  51.00  Sunday               1
3 2017-01-01 03:00:00  47.27  Sunday               1
4 2017-01-01 04:00:00  45.49  Sunday               1

In [8]:
dati['Giorno_Vacanza'].value_counts()

0    25416
1      864
Name: Giorno_Vacanza, dtype: int64

Corretto poichè ci sonp 12 giorni festivi spalamati su 24 ore e 3 anni. Quindi 12 * 24 * 3 = 864

In [9]:
dummies = pd.get_dummies(dati['Giorno'])

In [10]:
frames = [dati, dummies, sinusoidi]

In [11]:
dati = pd.concat(frames, axis = 1)

In [13]:
dati = dati.drop(columns = ['Giorno'])

In [14]:
dati.head()

Data    PUN  Giorno_Vacanza  Friday  Monday  Saturday  \
0 2017-01-01 00:00:00  53.30               1       0       0         0   
1 2017-01-01 01:00:00  52.00               1       0       0         0   
2 2017-01-01 02:00:00  51.00               1       0       0         0   
3 2017-01-01 03:00:00  47.27               1       0       0         0   
4 2017-01-01 04:00:00  45.49               1       0       0         0   

   Sunday  Thursday  Tuesday  Wednesday  ...  sinu365.23  sinu365.24  \
0       1         0        0          0  ...    0.248940   -0.160575   
1       1         0        0          0  ...   -0.482206    0.316982   
2       1         0        0          0  ...    0.685111   -0.465162   
3       1         0        0          0  ...   -0.844881    0.601271   
4       1         0        0          0  ...    0.951454   -0.721775   

   sinu365.25  sinu365.26  sinu365.27  sinu365.28  sinu365.29  sinu365.30  \
0   -0.543105   -0.834370   -0.985424   -0.970882   -0.793190   -0.482206   
1    0.912053    0.919817    0.335276   -0.465162   -0.966065   -0.844881   
2   -0.988531   -0.179645    0.871351    0.748017   -0.383428   -0.998123   
3    0.748017   -0.721775   -0.631740    0.823547    0.499069   -0.903946   
4   -0.267636    0.975336   -0.656411   -0.353445    0.991269   -0.585696   

   sinu365.31  sinu365.32  
0   -0.090190    0.316982  
1   -0.179645    0.601271  
2   -0.267636    0.823547  
3   -0.353445    0.960885  
4   -0.436373    0.999121  

[5 rows x 42 columns]